### ISSUE TITLE
Static Discovery Example

### ISSUE THREAD

**Issue by adasta** [eProsima/Fast-RTPS#98]
It would be good to supply a more concrete static discovery example.  This topic is [explained here](http://eprosima-fast-rtps.readthedocs.io/en/latest/advanced.html#matching-endpoints-the-manual-way).  However, It is not clear how all of the locator lists interact.  

My scenario is the following.   There are two participants on two different machines,  machine A is IP address ( 10.1.10.30) and machine B is IP address (10.1.10.31) .  Each machine's program has a publisher talking to the other computer as well as a subscriber receiving from the other computer.  

There seem to be 3 different locations that static routing can be specified.  It can be set in the ParticipantAttributes,  PublisherAttributes, and Subscriber attributes.  
  
1. ParticipantAttributes
  -  defaultMulticastLocatorList
  -  defaultOutLocatorList
  -  defaultUnicastLocatorList
2. SubscriberAttributes:
  -   multicastLocatorList
  -  outLocatorList
  -  UnicastLocatorList
3. PublisherAttributes:
  -   multicastLocatorList
  -  outLocatorList
  -  UnicastLocatorList

The documentation for all of these locator lists seem to indicate that  the locators above control which network interface and IP address the program will bind to on its computer.  They do not seem to specify the IP address and port of the remote peer to which the computer should speak.  

An adaptation of the HelloWorld program for static discovery would be greatly appreciated.

**Comment by JaimeMartin** [comment]
Hi Adam, 

I replied your email.

**Comment by ufechner7** [comment]
Any progress on this?
I have two machines, connected via a wireless on the same network segment using WDS, and automatic discovery is not working. I would like to tell the program (the publsher) on which IP address it can find the subscriber, but I don't now where.

**Comment by glkhobragade** [comment]
@JaimeMartin , @adasta , can you please share the xml file for the static discovery Hello world example.
It would be highly appreciated.

**Comment by glkhobragade** [comment]
The tutorial mentions in the below link seems not to be working.
http://eprosima-fast-rtps.readthedocs.io/en/latest/advanced.html#matching-endpoints-the-manual-way

**Comment by ufechner7** [comment]
Any update on this?

**Comment by richiware** [comment]
A HelloWorld example using static discovery was included in the repository ([link](https://github.com/eProsima/Fast-RTPS/tree/master/examples/C%2B%2B/StaticHelloWorldExample)).
Also there is a [XSD](https://github.com/eProsima/Fast-RTPS/blob/master/resources/xsd/fastRTPS_profiles.xsd) file to parse static discovery xml files.

**Comment by GlennCoombs** [comment]
The StaticHelloWorldExample only shows how to use static endpoint discovery but I would like to know how to do static participant discovery.  The original post by adasta linked to a page describing locator lists but it isn't obvious how to use them.  It mentions initialPeersList, metatrafficUnicastLocatorList, metatrafficMulticastLocatorList, defaultUnicastLocatorList and defaultMulticastLocatorList.  It appears that the metatraffic locators (and maybe the initialPeersList) are used during the participant discovery while the default lists are used by the user publishers/subscribers.

I would like to just hard code the IP addresses and ports used so that participant discovery is not used.  I have one publisher on 10.0.0.1 and four subscribers on IP addresses 10.0.1.5, 10.0.1.6, 10.0.1.7 and 10.0.1.8.  If I want to publish a single message and have all subscribers receive it what do I need to do ?

I'm guessing that I can set the initialPeersList and both metatraffic locators to empty as there won't be any metatraffic ?  Would I then set the defaultUnicastLocatorList to contain a single entry for 10.0.1.1 on a random port like 22222 on each of the subscribers ? And set the defaultUnicastLocatorList on the publisher to contain 4 entries, one for each of the IP addresses ?

Is there any example code similar to the StaticHelloWorldExample that shows how to dispense with the Participant Discovery Phase ?

**Comment by richiware** [comment]
Static discovery is only for the EDP (Endpoint Discovery Phase). There is not static discovery for participant right now.

We provide an [example](https://github.com/eProsima/Fast-RTPS/tree/master/examples/C%2B%2B/RTPSTest_registered) using the low RTPS API to make communication without discovery. I hope this is useful for you.

**Comment by MiguelCompany** [comment]
@GlennCoombs We hope the provided example was useful. Closing this now. Feel free to reopen this if necessary

**Comment by 100429061** [comment]
@JaimeMartin Im trying to use the static discovery. However, the examples provided are not enough to completely understand how it works. Can you share more info or examples about the static discovery?

**Comment by EduPonz** [comment]
Hi @100429061,

I'm leaving here a couple of documentation links that may help you:

1. [Static discovery documentation](https://fast-dds.docs.eprosima.com/en/latest/fastdds/discovery/static.html)
2. [Well known network deployments use case](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/well_known_deployments/well_known_deployments.html#well-known-deployments)

If you were to have more specific questions the best way to proceed would be to open a [Q&A discussion](https://github.com/eProsima/Fast-DDS/discussions/categories/q-a). That way, other users with similar inquiries may refer to that conversation for reference.

**Comment by 100429061** [comment]
Hi Eduardo,
I have already seen the links you sent me. On the second one, a xml file
with the profiles is defined but is not used in the c++ codes. Besides, in
the subscriber's c++ code the HelloWorldPublisher.xml is used, as well as
the DataWriter configuration is specified in the code instead of the
DataReader configuration. Speaking of the DataWriter configuration, in both
c++ codes there is a DataWriter configuration but is not used.
I need to use the static EDP to communicate differentes machines. I suppose
I have to specify the IP address and port of the other machine but there
are no examples using static discovery that use this method.
Regards,
Gabriel García

**Comment by 100429061** [comment]
Hi Eduardo,
I've finally done a simple implementation with the static EDP. I think the
documentation about well known network deployments should be revised, I
have found a few errors like data_writer qos is implemented in the
Subscriber and the Publisher c++ code and it is not used when the
DataWriter or the DataReader is created (the default qos is used).
Regards,
Gabriel García

**Comment by EduPonz** [comment]
Thanks a lot @100429061 , we'll take a look at it and update accordingly!

**Comment by guanghuinie** [comment]
Hi,  @EduPonz  @100429061 

   I am running a static configuration example on two machines. When I don't configure the unicastLocator, the communication between the writer and reader works fine. 
    However, when I add the unicastLocator configuration, Wireshark shows "Destination unreachable (Port unreachable)" error. I have tried multiple ports, and the issue persists, ruling out any port occupancy problems.
   When the unicastLocator was configured with port 3100, it showed a "Port unreachable" error. At that time, I observed the following message in the example output: "2022-09-10 20:25:38.326814 [RTPS_PARTICIPANT Error] 3672 HelloWorldPublisher RTPSParticipantImpl uses unicast, and the defaultUnicastLocatorList: 127.0.0.1:7413,172.20.1.22:7413,172.20.2.22:7413, -> Function createAndAssociateReceiverswithEndpoint." However, when I configured the unicastLocator with port 7413, the communication between the writer and reader worked fine.


Any suggestions on how to configure unicatLocator for static service discovery
